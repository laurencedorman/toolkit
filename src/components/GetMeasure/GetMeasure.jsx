/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import PropTypes from 'prop-types';
import useMeasure from './hook';

const GetMeasure = ({ children }) => {
  const [size, ref] = useMeasure();

  return children({ size, ref });
};

GetMeasure.displayName = 'GetMeasure';

GetMeasure.propTypes = {
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};

export default GetMeasure;
