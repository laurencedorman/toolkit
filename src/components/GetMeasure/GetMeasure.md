GetMeasure:

- Get the measure of component (Data are calculated with "getBoundingClientRect" method).
- All data are included in the component.
- Use as render child method (Check code of example).
- Just set a `ref={ref}` on target who need measure.
- Resize Observer included with polyfills.
- Scroll Listener included.

```jsx
import Layout from '../Layout';

<Layout direction="column" styleGuidist>
  <GetMeasure>
    {({ size, ref }) => (
      <>
        <div ref={ref}>
          <h3>Title component</h3>
          <p>
            Nunc yo arcu. Donec ante. We gonna chung placerizzle mauris at
            lectus. Morbi fo shizzle tortizzle. Crizzle yo mamma tortor nizzle
            mi consectetizzle sure. Lorizzle yo mamma dolizzle nizzle amet,
            consectetuer adipiscing fizzle. Ma nizzle fizzle arcu, things uhuh
            ... yih!, eleifend izzle, mollizzle in, est. Aliquizzle erat
            volutpizzle. Uhuh ... yih! tempizzle gangster turpis. Suspendisse
            aliquizzle. Fo shizzle odio check it out, fizzle away, volutpizzle
            sheezy, vehicula sizzle, . Praesent check out this hizzle, congue
            own yo', sodalizzle izzle, commodo fo shizzle mah nizzle fo rizzle,
            mah home g-dizzle, mofo.
          </p>
        </div>
        <br />
        <pre>Measure of this content: {JSON.stringify(size, null, 2)}</pre>
      </>
    )}
  </GetMeasure>
</Layout>;
```

Hook useMeasure

```jsx
import Layout from '../Layout';
import { useMeasure } from './index';

const [size, ref] = useMeasure();

<Layout direction="column" styleGuidist>
  <div ref={ref}>
    <h3>Title component</h3>
    <p>
      Nunc yo arcu. Donec ante. We gonna chung placerizzle mauris at lectus.
      Morbi fo shizzle tortizzle. Crizzle yo mamma tortor nizzle mi
      consectetizzle sure. Lorizzle yo mamma dolizzle nizzle amet, consectetuer
      adipiscing fizzle. Ma nizzle fizzle arcu, things uhuh ... yih!, eleifend
      izzle, mollizzle in, est. Aliquizzle erat volutpizzle. Uhuh ... yih!
      tempizzle gangster turpis. Suspendisse aliquizzle. Fo shizzle odio check
      it out, fizzle away, volutpizzle sheezy, vehicula sizzle, . Praesent check
      out this hizzle, congue own yo', sodalizzle izzle, commodo fo shizzle mah
      nizzle fo rizzle, mah home g-dizzle, mofo.
    </p>
  </div>
  <br />
  <pre>Measure of this content: {JSON.stringify(size, null, 2)}</pre>
</Layout>;
```
