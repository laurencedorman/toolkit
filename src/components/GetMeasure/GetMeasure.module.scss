/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

.wrapper {
  position: relative;
  width: 100%;
}
