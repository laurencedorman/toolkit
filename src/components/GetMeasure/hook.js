import { useLayoutEffect, useState, useRef } from 'react';
import ResizeObserver from 'resize-observer-polyfill';

export default () => {
  const ref = useRef(null);
  const [size, setSize] = useState({
    x: 0,
    y: 0,
    bottom: 0,
    height: 0,
    left: 0,
    right: 0,
    top: 0,
    width: 0,
  });

  useLayoutEffect(() => {
    const element = ref.current;
    const resizeObserver = new ResizeObserver(entries => {
      if (!Array.isArray(entries) || entries.length === 0) return;

      setSize(entries[0].target.getBoundingClientRect());
    });

    resizeObserver.observe(element);

    return () => resizeObserver.unobserve(element);
  }, []);

  return [size, ref];
};
