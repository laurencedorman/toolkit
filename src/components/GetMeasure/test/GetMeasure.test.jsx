/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { shallow } from 'enzyme';
import GetMeasure from '../GetMeasure';

jest.mock('../hook', () => {
  return () => [{ height: 300, width: 8000 }, { current: null }];
});

describe('GetMeasure', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<GetMeasure>{({ size }) => size}</GetMeasure>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass proper props to children', async () => {
    const children = jest.fn(({ ref }) => <div ref={ref}>Dumb component</div>);

    shallow(<GetMeasure>{children}</GetMeasure>);

    expect(children).toHaveBeenCalledWith({
      size: { width: 8000, height: 300 },
      ref: { current: null },
    });
  });
});
