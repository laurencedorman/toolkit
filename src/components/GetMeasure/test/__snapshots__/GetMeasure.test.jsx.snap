// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetMeasure should match snapshot 1`] = `<Component />`;
